### ISSUE TITLE
Question about tokenizer design

### ISSUE THREAD

**Issue by YusenHou519** [lingxusb/megaDNA#3]
（English version is attached）

邵教授您好，

非常感谢您在 megaDNA 上的出色工作，我受益匪浅。

我在阅读代码时注意到词表定义为 `['**', 'A', 'T', 'C', 'G', '#']`，想向您请教一下其中 `**` 和 `#` 这两个 token 分别起什么作用？

**在训练模型时，是否需要在每个序列的开头加上 `**`，并在结尾加上 `#`？**

在 [megaDNA_generate.ipynb](https://github.com/lingxusb/megaDNA/blob/main/notebook/megaDNA_generate.ipynb)中，您直接将 primer_DNA 输入模型而未添加 `**`（token id: 0），并在生成后使用 `#` 进行分割。这让我推测 `#` 可能被用作 `[eos]`（end of sequence）token。

```
primer_sequence = torch.tensor(np.random.choice(np.arange(1,5), PRIME_LENGTH)).long().to(device)[None,]
primer_DNA = ''.join(map(token2nucleotide, primer_sequence[0]))
print(f"Primer sequence: {primer_DNA}\n{'*' * 100}")

# Generate a sequence using the model
seq_tokenized = model.generate(primer_sequence, 
                               seq_len=context_length,
                               temperature=0.95, 
                               filter_thres=0.0) 
generated_sequence = ''.join(map(token2nucleotide, seq_tokenized.squeeze().cpu().int()))

# Split the generated sequence into contigs at the '#' character
contigs = generated_sequence.split('#')
```

相比之下，在 [megaDNA_mutagenesis.ipynb](https://github.com/lingxusb/megaDNA/blob/main/notebook/megaDNA_mutagenesis.ipynb)中，当计算序列的 loss 时，函数 encode_sequence 会在序列前加上 `**`，在末尾加上 `#`。这似乎表明 `**` 是 `[bos]`（beginning of sequence）token，而 `#` 是 `[eos]` token。

```
def encode_sequence(sequence, nt_vocab=nt):
  """Encode a DNA sequence to its numerical representation."""
  return [0] + [nt_vocab.index(nucleotide) if nucleotide in nt_vocab else 1 for nucleotide in sequence] + [5]
```

不知道我的理解是否正确，期待您的解答。

非常感谢您的时间！

祝好，
Kiikii

Hi Prof. Shao,

Thank you very much for your excellent work on megaDNA! I really enjoyed reading it.

I have a few questions regarding the tokenizer implementation. After reading the code carefully, I noticed that the vocabulary is defined as `['**', 'A', 'T', 'C', 'G', '#']`.
I am wondering what are the purposes of the `**` and `#` tokens.

**When training the model, should I manually add `**` at the beginning and `#` at the end of each sequence?**

In [megaDNA_generate.ipynb](https://github.com/lingxusb/megaDNA/blob/main/notebook/megaDNA_generate.ipynb), the `primer_DNA` is directly passed into the model without adding `**` (token id: 0). The generated sequences are later split using `#`. This makes me think that `#` might serve as an end-of-sequence (`[eos]`) token.

```
primer_sequence = torch.tensor(np.random.choice(np.arange(1,5), PRIME_LENGTH)).long().to(device)[None,]
primer_DNA = ''.join(map(token2nucleotide, primer_sequence[0]))
print(f"Primer sequence: {primer_DNA}\n{'*' * 100}")

# Generate a sequence using the model
seq_tokenized = model.generate(primer_sequence, 
                               seq_len=context_length,
                               temperature=0.95, 
                               filter_thres=0.0) 
generated_sequence = ''.join(map(token2nucleotide, seq_tokenized.squeeze().cpu().int()))

# Split the generated sequence into contigs at the '#' character
contigs = generated_sequence.split('#')
```
In contrast, in [megaDNA_mutagenesis.ipynb](https://github.com/lingxusb/megaDNA/blob/main/notebook/megaDNA_mutagenesis.ipynb)
, when computing the loss for a given sequence, the function `encode_sequence` adds `**` at the beginning and `#` at the end. This suggests that `**` might represent the beginning-of-sequence (`[bos]`) token, and `#` the `[eos]` token.
```
def encode_sequence(sequence, nt_vocab=nt):
  """Encode a DNA sequence to its numerical representation."""
  return [0] + [nt_vocab.index(nucleotide) if nucleotide in nt_vocab else 1 for nucleotide in sequence] + [5]
```

Could you please confirm whether this interpretation is correct?
And when training a new model from scratch, should we explicitly include these tokens in the input sequences?

Thank you very much for your time and clarification!

Best regards,
Kiikii

**Comment by lingxusb** [comment]
Hi Yusen, thank you for your excellent questions! Yes, `#` is the `[eos]` token. I recommend including the `[eos]` token during training, and you’ll also need it when generating new sequences.

Regarding the `**` token, I don't think adding it at the beginning during training will make much difference, since it will work just as a soft prompt. I hope this answers your question. I appreciate your careful reading of our codes and let me know if you have any further questions. Thanks!

Best,
Bin

**Comment by YusenHou519** [comment]
Dear Prof. Shao,

Thanks a lot for your detailed and informative response. It helps me better understand the models' training process!
Really enjoy your excellent work! Really excited to see more wonderful works from your lab.

Best regards,
Yusen
